src/optimizer/binder.py

```diff
@@ -0,0 +1,74 @@
+# coding=utf-8
+# Copyright 2018-2020 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import itertools
+
+from src.optimizer.rules.pattern import Pattern
+from src.optimizer.group_expression import GroupExpression
+from src.optimizer.memo import Memo
+from src.optimizer.operators import Operator, OperatorType
+
+
+class Binder:
+    def __init__(self, grp_expr: GroupExpression, pattern: Pattern, memo: Memo):
+        self._grp_expr = grp_expr
+        self._pattern = pattern
+        self._memo = memo
+
+    @staticmethod
+    def _grp_binder(id: int, pattern: Pattern, memo: Memo):
+        grp = memo.groups[id]
+        grp_exprs = grp.logical_exprs
+
+        for expr in grp.logical_exprs:
+            yield from Binder._binder(expr, pattern, memo)
+
+    @staticmethod
+    def _binder(expr: GroupExpression, pattern: Pattern, memo: Memo):
+        curr_iterator = iter([expr.opr])
+        child_binders = []
+        if pattern.opr_type is not OperatorType.DUMMY:
+            if expr.opr.opr_type != pattern.opr_type:
+                return
+
+            if len(pattern.children) != len(expr.children):
+                return
+
+            for child_grp, pattern_child in zip(expr.children,
+                                                pattern.children):
+                child_binders.append(
+                    Binder._grp_binder(child_grp, pattern_child, memo)
+                )
+
+        yield from itertools.product(curr_iterator, *child_binders)
+
+    @staticmethod
+    def build_opr_tree_from_inorder_repr(inorder_repr: tuple) -> Operator:
+        if isinstance(inorder_repr, Operator):
+            return inorder_repr
+        else:
+            opr_tree = inorder_repr[0]
+            if len(inorder_repr) > 1:
+                # remove old children
```

Comment:
@xzdandy I am going to work on adding unittest for binder. I am little confused here. 
inorder_repr is the inorder traversal of the tree? If it is inorder, the left child will actually appear before the root right? 
Why do we remove the old children here? To prevent infinite loop?

Reply:
The group expression keeps the correct parent-child relationship between operators. So we need to clear the operator's original children list and use the group expression to generate the new children list.

Reply:
I believe it should be post order. The name of the function is wrong. @gaurav274 Correct me if I am wrong.

Reply:
Yeah, I always get confused by these order traversals. I think it should be pre-order. It is root followed by children from left to right.